Enable use of Mbed TLS as a CMake subproject

Patater opened the pull request:
Remove use of CMAKE_SOURCE_DIR in case mbedtls is built from within
another CMake project. Define MBEDTLS_DIR to ${CMAKE_CURRENT_SOURCE_DIR}
in the main CMakeLists.txt file and refer to that when defining target
include paths to enable mbedtls to be built as a sub project.

Fixes #2609

Signed-off-by: Ashley Duncan <ashes.man@gmail.com>
Signed-off-by: Jaeden Amero <jaeden.amero@arm.com>

## Description
A few sentences describing the overall goals of the pull request's commits.


## Status
**IN DEVELOPMENT**

* Will need to point to a crypto submodule containing https://github.com/ARMmbed/mbed-crypto/pull/152 on the mbed-crypto/development branch

## Requires Backporting
NO  - but may be useful in a 2.18.1

## Migrations
NO

## Additional comments
Any additional information that could be of interest

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

Patater commented:
Rebased to add information about this newly supported feature to the README and to add a regression test.

Patater commented:
Force pushed to update crypto submodule

Patater commented:
Rebased to update crypto submodule

Patater commented:
Rebased to not require too new of a CMake version.

andresag01 commented on tests/scripts/all.sh:
```diff
@@ -1262,6 +1262,25 @@ component_test_cmake_out_of_source () {
+    MBEDTLS_ROOT_DIR="$PWD"
+
+    # Remove build artifacts if any from previous builds.
+    rm -rf programs/test/cmake_subproject/build
```
Suggestion: Perhaps is worth adding this to the `cleanup()` instead of here?

Patater replied:
This step is generally not needed. Only this tests makes a mess there, so only this one cleans up its mess. That was my thinking.

andresag01 replied:
I do not feel very strongly about it, so we can leave it like that.

My suggestion was to ensure that this cleanup is executed regardless if there is an early failure in the test. For example, if we were to leave it here and a subsequent command before line 1238 fails, then the script might terminate and not do the cleanup step in line 1238. In contrast, I would expect the cleanup() call to be done upon exit

Patater replied:
OK, moved into `cleanup()`.

andresag01 commented on tests/scripts/all.sh:
```diff
@@ -1262,6 +1262,25 @@ component_test_cmake_out_of_source () {
+    unset MBEDTLS_ROOT_DIR
+
+    # Remove build artifacts if any.
+    rm -rf programs/test/cmake_subproject/build
```
Suggestion: Perhaps is worth adding this to the `cleanup()` instead of here?

Patater commented:
Rebased to change minimum CMake version to 2.6, to match the version required by the top level project `CMakeLists.txt`. Also updated the crypto submodule to the latest version of https://github.com/ARMmbed/mbed-crypto/pull/152

Patater commented:
Rebased to rebase atop latest development and resolve conflicts in ChangeLog and crypto submodule update.

Patater commented:
Rebased to move cleanup of the CMake subproject test to `cleanup()`, and to update to the latest https://github.com/ARMmbed/mbed-crypto/pull/152

Patater commented:
Rebased to update to the latest https://github.com/ARMmbed/mbed-crypto/pull/152

Patater commented:
Rebased to update Mbed Crypto to a commit on its development branch containing the merged https://github.com/ARMmbed/mbed-crypto/pull/152 PR.

jenia81 commented:
@Patater when is this PR expected to be merged?
We need the PSA multiple key usage feature - https://github.com/ARMmbed/mbedtls-psa/pull/266 that is part of Mbed TLS 2.18. And currently Mbed TLS 2.18 isn't compiled with PDMC on Linux, so this PR is needed
